### PR TITLE
feat: Support configuring annotations that are preserved by the AppSet (#12507)

### LINKED ISSUE #12507: ApplicationSet: Allow preserving state of custom annotations
# Summary

An ApplicationSet should be able to be configured to allow preserving the state of custom annotations. This is already hard-coded for the Argo CD Notifications and the Argo CD refresh type annotations, the reason this was done for those can also apply to custom annotations.

# Motivation

I'm using a custom plugin that patches the `Application` being refreshed with some context that is only available at refresh time. I need that context to be available to Argo CD Notifications (and other external systems) so I am storing that context in a custom annotation. The problem now is that the ApplicationSet controller will remove the annotation as it isn't defined on the `ApplicationSet`. Its a common pattern for operators and such to use custom annotations to keep some state, as seen by the annotations Argo CD itself uses.

# Proposal

I propose adding a new field to the `ApplicationSet` that allows users to configure annotations that should be preserved. This would look something like: `preservedFields.annotations: ["custom-anno"]`. This structure can be evolved in the future to support preserving other fields (labels could be a good candidate), though you may want to avoid allowing arbitrary fields to be preserved.

### PR DESCRIPTION
**Pull request by lacarvalho91**
Closes #12507

Note on DCO:

If the DCO action in the integration test fails, one or more of your commits are not signed off. Please click on the *Details* link next to the DCO action for instructions on how to resolve this.

Checklist:

* [X] Either (a) I've created an [enhancement proposal](https://github.com/argoproj/argo-cd/issues/new/choose) and discussed it with the community, (b) this is a bug fix, or (c) this does not need to be in the release notes.
* [X] The title of the PR states what changed and the related issues number (used for the release note).
* [x] I've included "Closes [ISSUE #]" or "Fixes [ISSUE #]" in the description to automatically close the associated issue.
* [ ] I've updated both the CLI and UI to expose my feature, or I plan to submit a second PR with them.
* [x] Does this PR require documentation updates?
* [x] I've updated documentation as required by this PR.
* [ ] Optional. My organization is added to USERS.md.
* [x] I have signed off all my commits as required by [DCO](https://github.com/argoproj/argoproj/tree/master/community#contributing-to-argo)
* [X] I have written unit and/or e2e tests for my change. PRs without these are unlikely to be merged.
* [x] My build is green ([troubleshooting builds](https://argo-cd.readthedocs.io/en/latest/developer-guide/ci/)).

### CODE REVIEW

File: pkg/apis/application/v1alpha1/applicationset_types.go [hunk 1]
```diff
@@ -54,11 +54,16 @@ func (a *ApplicationSet) RBACName() string {
...
-	Generators []ApplicationSetGenerator `json:"generators" protobuf:"bytes,2,name=generators"`
-	Template   ApplicationSetTemplate    `json:"template" protobuf:"bytes,3,name=template"`
-	SyncPolicy *ApplicationSetSyncPolicy `json:"syncPolicy,omitempty" protobuf:"bytes,4,name=syncPolicy"`
-	Strategy   *ApplicationSetStrategy   `json:"strategy,omitempty" protobuf:"bytes,5,opt,name=strategy"`
+	GoTemplate      bool                        `json:"goTemplate,omitempty" protobuf:"bytes,1,name=goTemplate"`
+	Generators      []ApplicationSetGenerator   `json:"generators" protobuf:"bytes,2,name=generators"`
+	Template        ApplicationSetTemplate      `json:"template" protobuf:"bytes,3,name=template"`
+	SyncPolicy      *ApplicationSetSyncPolicy   `json:"syncPolicy,omitempty" protobuf:"bytes,4,name=syncPolicy"`
+	Strategy        *ApplicationSetStrategy     `json:"strategy,omitempty" protobuf:"bytes,5,opt,name=strategy"`
+	PreservedFields *ApplicationPreservedFields `json:"preservedFields,omitempty" protobuf:"bytes,6,opt,name=preservedFields"`
```
Comment (lacarvalho91) [inline pkg/apis/application/v1alpha1/applicationset_types.go:62]: also considered `IgnoredFields` as the name, but the word preserved was already being used when preserving Argo CD notifications annotation and the refresh type one

**Comment by lacarvalho91** [comment]
I've been rebasing this against master to keep it up-to-date but I'll wait for a review before I do it again as I've done it a few times now

**Comment by lacarvalho91** [comment]
I had originally linked this PR to #9101 but after rereading the problem described on that issue I don't believe preserving annotations is enough to accommodate that use-case. Instead I have created a separate issue and linked this PR to that.

**Comment by lacarvalho91** [comment]
@crenshaw-dev any idea when this could be reviewed? any rough guess would help just so i can plan a bit :slightly_smiling_face:

**Comment by lacarvalho91** [comment]
@ishitasequeira would you be able to review please? :slightly_smiling_face:

**Review by crenshaw-dev — APPROVED** [review]

**Pull request merged**